### PR TITLE
Improved classifier

### PR DESCRIPTION
**Pull request by tomkocmi**
Clean PR of previous attempt. Solves all PR requests.

### CODE REVIEW

File: neuralmonkey/config/builder.py [hunk 1]
```diff
@@ -26,21 +26,30 @@ def create(self):
...
-            # if the problem is really importing the module
-            if exc.name == module_name:
-                raise Exception(
-                    ("Interpretation '{}' as type name, module '{}' "
-                     "does not exist. Did you mean file './{}'? \n{}")
-                    .format(self.clazz,
-                            module_name, self.clazz, exc)) from None
-            else:
-                raise
+            import tensorflow as tf
```
Comment (jlibovicky) [inline neuralmonkey/config/builder.py]: Proč se to importuje zrovna tady? Nebylo by lepší to dát úplně nahoru, když až na výjimky většina jmen tímhle projde?

Comment (jindrahelcl) [inline neuralmonkey/config/builder.py]: hlavně je to v `try` bloku, kterej chytá `ImportError`, a ten `ImportError` můžou vyhodit oba dva řádky

File: neuralmonkey/config/builder.py [hunk 2]
```diff
@@ -26,21 +26,30 @@ def create(self):
...
+            try:
+                if class_parts[0] == 'tf':
+                    # Due to the architecture of TensorFlow, it must be
+                    # imported this way.
+                    tf = importlib.import_module("tensorflow")
+                    module = getattr(tf, ".".join(class_parts[1:-1]))
+                else:
+                    module_name = ".".join(["neuralmonkey"] + class_parts[:-1])
+                    module = importlib.import_module(module_name)
+            except ImportError as exc:
```
Comment (jindrahelcl) [inline neuralmonkey/config/builder.py]: tohle mi přijde hodně divný, chytat `ImportError` v `except` bloku, kterej se raisne při import erroru. A předtim nefungovalo v konfiguraci tf? Podle mě by mělo fungovat, když se dá `import tensorflow as tf` a potom se volá `import_module` na modul z konfigurace.

A když ne, tak bych udělal pár regexpovejch pravidel, který budou preprocessovat vstup z konfiguráku (tedy `^tf\.` to nahradí za `tensorflow` a je to).

File: neuralmonkey/decoders/output_projection.py [hunk 3]
```diff
@@ -5,7 +5,7 @@
 import tensorflow as tf
 
 from neuralmonkey.nn.projection import maxout
-from neuralmonkey.nn.mlp import multilayer_perceptron
+from neuralmonkey.nn.projection import multilayer_projection
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/output_projection.py]: proč? Vždyť to je multilayer perceptron..

File: neuralmonkey/decoders/sequence_classifier.py [hunk 4]
```diff
@@ -23,18 +23,33 @@ def __init__(self,
                  encoders: List[Any],
                  vocabulary: Vocabulary,
                  data_id: str,
-                 layers: Optional[List[int]]=None,
-                 activation: Callable[[tf.Tensor], tf.Tensor]=tf.tanh,
+                 layers: List[int]=None,
+                 activation_fn: Callable[[tf.Tensor], tf.Tensor]=tf.nn.relu,
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/sequence_classifier.py]: nikdo to nepoužívá, takže tady OK, ale jinak prosím neměnit default parametry. Špatně se toho všímá a už jsem takhle jednou vytuhnul na dva měsíce s experimentama, který najednou přestaly běhat.

File: neuralmonkey/decoders/sequence_classifier.py [hunk 5]
```diff
@@ -23,18 +23,33 @@ def __init__(self,
...
                  dropout_keep_prob: float=0.5,
                  save_checkpoint: Optional[str]=None,
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/sequence_classifier.py]: když uhryzáváš `Optional`, tak to dělej konzistentně, jinak to akorát zanáší bordel. Já osobně nevim, jak to přesně funguje, mám za to, že je to spíš zbytečný, když tomu dáváš None jako default hodnotu. Ale když budeš mít typ `List[str]=None`  a pak se v tý funkci změní default na něco jinýho a None zůstane přípustná hodnota, tak tam už to Optional bejt má, takže by asi bylo chytrý to tam nechávat všude.

File: neuralmonkey/nn/projection.py [hunk 6]
```diff
@@ -19,9 +19,12 @@ def linear(inputs, size, scope="LinearProjection"):
         A tensor of shape batch x size
     """
     with tf.variable_scope(scope):
-        return tf.nn.seq2seq.linear(inputs, size, True)
+        # TODO this will be soon deprecated, should be rewrited with
+        # tf.contrib.layers.fully_connected and a support for list of tensors
```
Comment (jindrahelcl) [inline neuralmonkey/nn/projection.py]: jestli `tf.contrib.layers.fully_connected` neumí list of tensors, tak si tu projekci uděláme tady sami. Podle mě neni důvod to přebírat z contrib, protože to je natolikk jednoduchá operace, že se vyplatí mít to tady, aby se každej mohl podívat na ten kód.. Už jsi viděl `tf.contrib.layers.fully_connected`?

File: neuralmonkey/nn/projection.py [hunk 7]
```diff
@@ -19,9 +19,12 @@ def linear(inputs, size, scope="LinearProjection"):
...
+        return tf.nn.seq2seq.linear(inputs, size, True, scope=scope)
 
 
+# SUGGESTION: what about using ReLu as a default? It have better results
```
Comment (jindrahelcl) [inline neuralmonkey/nn/projection.py]: Proč by to měl kvůli lepším výsledkům být default? Tady bych spíš vůbec žádný default nevyplňoval.

Comment (jindrahelcl) [inline neuralmonkey/nn/projection.py]: To je nevhodně formovaná otázka - chci říct, že žiju v domění, že v knihovnách je zvykem mít u těchto věcí `tanh` jako default, takže dávat tam něco jinýho by mohlo být zdrojem problémů spojených s reprodukcí :-)

File: neuralmonkey/nn/projection.py [hunk 8]
```diff
@@ -64,3 +67,17 @@ def maxout(inputs, size, scope="MaxoutProjection"):
 
         reshaped = tf.reshape(maxpooled, [-1, size])
         return reshaped
+
+
+def multilayer_projection(input_, layer_sizes, activation=tf.nn.relu,
```
Comment (jindrahelcl) [inline neuralmonkey/nn/projection.py]: sice je to projekce, ale je to perceptron a patří to podle mě jinam. Taky tenhle kód dělá to samý co ta třída, takže dávat to jinam ještě víc zamaskuje, že máme duplicitní kód.

Btw. pořád jsem přesvědčen, že tahle funkce je to, co by se mělo používat (a ne ta třída)

File: neuralmonkey/config/builder.py (hunk 1 above)
Comment (tomkocmi) [inline neuralmonkey/config/builder.py]: pardon, tohle mi zde zbyl debugovaci kod, byl s tim dost problem to udelat

File: neuralmonkey/config/builder.py (hunk 2 above)
Comment (tomkocmi) [inline neuralmonkey/config/builder.py]: Tuto variantu jsme s @jlibovicky resili, bohuzel to opravdu nejde, protoze importovani tf neni klasicke importovani a neexistuje napriklad nic jako tf.nn.relu ... misto toho tim jak importujes tensorflow tak ve skutecnosti importujes pouze prepisy adres kde treba zrovna to tensorflow.nn se prepise na tensorflow.python.ops.nn ... toto je nejcistsi a funkcni varianta

File: neuralmonkey/decoders/output_projection.py (hunk 3 above)
Comment (tomkocmi) [inline neuralmonkey/decoders/output_projection.py]: neni, nema to vystupni vrstvu, viz @jlibovicky

File: neuralmonkey/decoders/sequence_classifier.py (hunk 4 above)
Comment (tomkocmi) [inline neuralmonkey/decoders/sequence_classifier.py]: no puvodne jsem jej chtel zmenit i v nelinearni projekci, ale tu jsem prave videl ze se pouziva na plno mistech, tak jsem jen vlozil koment, ze by to stalo za to dat jinej default (preci jen relu docela valcuje tanh, napric ruznymi clanky). Vic by k tomu mozna vedel @foxik :)

File: neuralmonkey/decoders/sequence_classifier.py (hunk 5 above)
Comment (tomkocmi) [inline neuralmonkey/decoders/sequence_classifier.py]: Ty jo, tak tohle optional absolutne netusim jak ma fungovat. Co jsem si k tomu nasel, tak jsem pochopil ze to z toho parametru dela dobrovolny parametr, coz ale pak nefunguje, takze jsem to zrusil a to none mi tam asi zbylo omylem, tak jej take zrusim. Ty layers jsou proste povinny parametr v teto tride.

File: neuralmonkey/nn/projection.py (hunk 7 above)
Comment (tomkocmi) [inline neuralmonkey/nn/projection.py]: Ja si myslim, ze stejne to nepujde poradne 100% schopne reprodukovat a pak naopak je lepsi, kdyz tam bude neco, co se chova lepe nez horsi default. Ale mozna to je muj nazor, kdy uprednostnuju lepsi vysledky nad porovnatelnosti s jinyma frameworkama. Reseni by asi bylo nedavat tam zadny default, ale to podle me rozbije dost veci.

File: neuralmonkey/nn/projection.py (hunk 8 above)
Comment (tomkocmi) [inline neuralmonkey/nn/projection.py]: tahle funkce dela neco jineho, nema vystupni vrstvu. Resil jsem to vcera s @jlibovicky

**Comment by tomkocmi** [comment]
Bohuzel nevim jak se vyrabi testovaci konfiguraky a jak se s nimi pracuje. Byl by to urcite dobry napad, ale osobne netusim jaky v nich mate system, jak se zapojuji atp.

File: neuralmonkey/decoders/output_projection.py (hunk 3 above)
Comment (jlibovicky) [inline neuralmonkey/decoders/output_projection.py]: Multilayer perceptron je klasifikátor, kde je několik nelineárních vrstev a na konci softmax. Tohle je projekce. Třída `MultiLayerPerceptron` tuhhle funkci využívá a navrch přidá softmax.

File: neuralmonkey/nn/projection.py (hunk 6 above)
Comment (jlibovicky) [inline neuralmonkey/nn/projection.py]: Podle mě je jedno, co tam bude za default, protože hodně záleží na aplikaci, co je lepší. _ReLU_ bezpochyby dává lepší výsledky u mnohovrstvého klasifikátru. _tanh_ je zase vhodný tam, kde je potřeba jenom jedna nelinearita nebo je to součástí nějakého složitějšího mechanismu, třeba uvnitř attentionu.

**Comment by jlibovicky** [comment]
Ten test case se udělá analogicky jako třeba `small.ini`. Vezmeš malý sample dat, konfigurák, co pro tu klasifikaci už tak jako tak používáš, nastavíš miniaturní velikosti všech skytých vrstev a trénování tak dlouho, aby aspoň jednou proběhla validace. Mělo by tam taky někde být to `tf.nn`, protože to vypadá docela křehce, aby se to testovalo.

**Comment by tomkocmi** [comment]
Pridal jsem tedy test a otestoval ... bezi zhruba 10 sekund a probehnou dve validace

File: neuralmonkey/nn/projection.py (hunk 8 above)
Comment (jindrahelcl) [inline neuralmonkey/nn/projection.py]: a nedalo by se to tedy refaktorovat tak, aby ta třída používala tuhle funkci a nad ní dala tu výstupní vrstvu?

File: neuralmonkey/nn/projection.py (hunk 7 above)
Comment (jindrahelcl) [inline neuralmonkey/nn/projection.py]: No nemělo by to rozbít nic, tohle je hluboko privátního API a odevšad se to volá s argumentem. Asi bych to nejradši vyhodil. Zkus to, a podle mě testy projdou.

File: neuralmonkey/nn/projection.py (hunk 8 above)
Comment (jlibovicky) [inline neuralmonkey/nn/projection.py]: Achjo, včera jsem to sáhodlouze komentoval a teď tady ten komentář vůbec nevidím. MLP je klasifikátor, tj. má na konci softmax. Tohle je jenom projekce. Třída MLP už teď tuhle funkci volá.

File: neuralmonkey/nn/projection.py (hunk 7 above)
Comment (jlibovicky) [inline neuralmonkey/nn/projection.py]: Nedá se jasně říct, že ReLU je lepší. ReLU je lepší, když je uvnitř vícevrstvého klasifikátoru. Pokud je to nelinearity kdesi v hloubi nějakého modelu, typicky se používá tanh - třeba při výpočtu aktivací attentionu.

File: tests/classifier.ini [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+[decoder_vocabulary]
+class=vocabulary.from_wordlist
+path="tests/data/classification.vocab"
+
+[decoder]
+class=decoders.sequence_classifier.SequenceClassifier
+name="decoder"
+encoders=[<encoder>]
+dropout_keep_prob=0.5
+layers: [30,10]
```
Comment (jlibovicky) [inline tests/classifier.ini]: Nemá tam být `=` místo `:`? To nespadne?

File: tests/classifier.ini [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+batch_size=16
+epochs=20
+train_dataset=<train_data>
+val_dataset=<val_data>
+trainer=<trainer>
+runners=[<runner>]
+postprocess=None
+evaluation=[("classification", <accuracy>)]
+logging_period=200
+validation_period=600
```
Comment (jlibovicky) [inline tests/classifier.ini]: Epochs, logging_period, validation_period, klidně se všechno může smrsknout na čtvrtinu.

File: tests/classifier.ini [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+[encoder_vocabulary]
+class=vocabulary.from_file
+path="tests/tmp-vocab/encoder_vocab.pkl"
+
+[encoder]
+class=encoders.sentence_encoder.SentenceEncoder
+name="sentence_encoder"
+rnn_size=7
+max_input_len=10
```
Comment (jlibovicky) [inline tests/classifier.ini]: To může být klidně třeba 4, stejně ten vstup má vždycky velikost 1.

File: tests/classifier.ini (hunk 9 above)
Comment (tomkocmi) [inline tests/classifier.ini]: fakt ze nespadlo, prijde mi to divny, ale beha to spravne s rovnitkem i dvojteckou

Comment (tomkocmi) [inline tests/classifier.ini]: ve vypisu to spravne pise nastavene velikosti, takze to funguje i s rovnitkem i dvojteckou

File: tests/classifier.ini [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+[val_data]
+class=dataset.load_dataset_from_files
+s_source="tests/data/val.tc.en"
+s_classification="tests/data/val.words"
+preprocessors=[("source", "source_chars", processors.helpers.preprocess_char_based)]
+
+[encoder_vocabulary]
+class=vocabulary.from_file
+path="tests/tmp-vocab/encoder_vocab.pkl"
```
Comment (jindrahelcl) [inline tests/classifier.ini]: jsi si jistej timhle? Bude to slovník, kterej se vytvořil ve `vocab.ini` a bude to tim pádem potřeba vždy pouštět až pozdějc.

File: neuralmonkey/config/builder.py [hunk 13]
```diff
@@ -26,21 +26,29 @@ def create(self):
...
+                # if the problem is really importing the module
+                if exc.name == module_name:
+                    raise Exception(
+                        ("Interpretation '{}' as type name, module '{}' "
+                         "does not exist. Did you mean file './{}'? \n{}")
+                        .format(self.clazz, module_name, self.clazz,
```
Comment (jindrahelcl) [inline neuralmonkey/config/builder.py]: Tahle hláška by se měla změnit, protože ty rady s tim souborem už neplatí. Mělo by se to zjednodušit prostě na cannot import module {}.format(module)
